### PR TITLE
Add support for Argyris, Bell, Hermite, and Morley elements

### PR DESCRIPTION
**Pull request by wence-**
Mostly tests, just need to adapt to new kernel interface that additional can provide a vertex-discretised local cell size for basis function rescaling.

### CODE REVIEW

**Comment by rckirby** [comment]
This also disallows Dirichlet BC on the new elements since we don't know
how to support them yet.

File: firedrake/assemble.py [hunk 1]
```diff
@@ -393,7 +393,7 @@ def thunk(bcs):
             loops.append(zero_tensor)
         else:
             zero_tensor()
-        for indices, (kernel, integral_type, needs_orientations, subdomain_id, domain_number, coeff_map, needs_cell_facets, pass_layer_arg) in kernels:
+        for indices, (kernel, integral_type, needs_orientations, subdomain_id, domain_number, coeff_map, needs_cell_facets, pass_layer_arg, needs_cell_sizes) in kernels:
```
Comment (miklos1) [inline firedrake/assemble.py]: This is now really begging for something more descriptive than an n-tuple... (`namedtuple`?)

File: firedrake/bcs.py [hunk 2]
```diff
@@ -43,6 +43,11 @@ class DirichletBC(object):
     '''
 
     def __init__(self, V, g, sub_domain, method="topological"):
+        # First, we bail out on zany elements.  We don't know how to do BC's for them.
+        import finat
+        if isinstance(V.finat_element, (finat.Hermite, finat.Argyris, finat.Morley)):
```
Comment (miklos1) [inline firedrake/bcs.py]: Bell is not zany?

File: firedrake/mesh.py [hunk 3]
```diff
@@ -1032,6 +1032,27 @@ def coordinates(self, value):
 
         raise AttributeError(message)
 
+    @utils.cached_property
+    def cell_sizes(self):
+        """A :class`~.Function` in the :math:`P^1` space containing the local mesh size.
+
+        This is computed by the :math:`L^2` projection of the local mesh element size."""
```
Comment (miklos1) [inline firedrake/mesh.py]: Thanks, this now clarifies the P1 mesh size, and why that needs to pre-calculated like `cell_orientations`...

File: tests/regression/test_helmholtz_zany.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+    # Analytical solution
+    f.project(Expression("cos(x[0]*pi*2)*cos(x[1]*pi*2)"))
+    return sqrt(assemble(dot(x - f, x - f) * dx)), x, f
+
+
+@pytest.mark.parametrize(('el', 'deg', 'convrate'),
+                         [('Hermite', 3, 3.8),
+                          ('Bell', 5, 4.8),
+                          ('Argyris', 5, 5)])
```
Comment (miklos1) [inline tests/regression/test_helmholtz_zany.py]: Morley not good for Helmholtz?

**Comment by rckirby** [comment]
On Tue, Jul 3, 2018 at 11:07 AM, Miklós Homolya <notifications@github.com>
wrote:

> *@miklos1* approved this pull request.
> ------------------------------
>
> In firedrake/assemble.py
> <https://github.com/firedrakeproject/firedrake/pull/1245#discussion_r199863311>
> :
>
> > @@ -393,7 +393,7 @@ def thunk(bcs):
>              loops.append(zero_tensor)
>          else:
>              zero_tensor()
> -        for indices, (kernel, integral_type, needs_orientations, subdomain_id, domain_number, coeff_map, needs_cell_facets, pass_layer_arg) in kernels:
> +        for indices, (kernel, integral_type, needs_orientations, subdomain_id, domain_number, coeff_map, needs_cell_facets, pass_layer_arg, needs_cell_sizes) in kernels:
>
> This is now really begging for something more descriptive than an
> n-tuple... (namedtuple?)
>
@wence?

> ------------------------------
>
> In firedrake/bcs.py
> <https://github.com/firedrakeproject/firedrake/pull/1245#discussion_r199863488>
> :
>
> > @@ -43,6 +43,11 @@ class DirichletBC(object):
>      '''
>
>      def __init__(self, V, g, sub_domain, method="topological"):
> +        # First, we bail out on zany elements.  We don't know how to do BC's for them.
> +        import finat
> +        if isinstance(V.finat_element, (finat.Hermite, finat.Argyris, finat.Morley)):
>
> Bell is not zany?
>

Good catch.


> ------------------------------
>
> In firedrake/mesh.py
> <https://github.com/firedrakeproject/firedrake/pull/1245#discussion_r199863742>
> :
>
> > @@ -1032,6 +1032,27 @@ def coordinates(self, value):
>
>          raise AttributeError(message)
>
> +    @utils.cached_property
> +    def cell_sizes(self):
> +        """A :class`~.Function` in the :math:`P^1` space containing the local mesh size.
> +
> +        This is computed by the :math:`L^2` projection of the local mesh element size."""
>
> Thanks, this now clarifies the P1 mesh size, and why that needs to
> pre-calculated like cell_orientations...
> ------------------------------
>
> In tests/regression/test_helmholtz_zany.py
> <https://github.com/firedrakeproject/firedrake/pull/1245#discussion_r199864237>
> :
>
> > +
> +    # Compute solution
> +    assemble(a)
> +    assemble(L)
> +    x = Function(V)
> +    solve(a == L, x, solver_parameters={'ksp_type': 'preonly', 'pc_type': 'lu'})
> +
> +    # Analytical solution
> +    f.project(Expression("cos(x[0]*pi*2)*cos(x[1]*pi*2)"))
> +    return sqrt(assemble(dot(x - f, x - f) * dx)), x, f
> +
> +
> +@pytest.mark.parametrize(('el', 'deg', 'convrate'),
> +                         [('Hermite', 3, 3.8),
> +                          ('Bell', 5, 4.8),
> +                          ('Argyris', 5, 5)])
>
> Morley not good for Helmholtz?
>

No, it is not a suitable C^0 nonconforming element.  Comment added.

File: tests/regression/test_helmholtz_zany.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+import numpy as np
+import pytest
+
+from firedrake import *
+
+
+def helmholtz(x, el_type, degree, mesh=None):
+    # Create mesh and define function space
+    if mesh is None:
+        mesh = UnitSquareMesh(2 ** x, 2 ** x)
```
Comment (miklos1) [inline tests/regression/test_helmholtz_zany.py]: Would it make sense to also test these elements on less regular meshes?

File: firedrake/bcs.py (hunk 2 above)
Comment (wence-) [reply firedrake/bcs.py]: Oh yeah, should add that.

File: firedrake/assemble.py (hunk 1 above)
Comment (wence-) [reply firedrake/assemble.py]: So in tsfc_interface it is a named tuple, I should destructure by name here.

File: firedrake/mesh.py (hunk 3 above)
Comment (wence-) [reply firedrake/mesh.py]: Yes, I'll add some explanatory comments on the tsfc side.

**Pull request merged**